ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala

```diff
@@ -329,29 +331,59 @@ private class PostgresLedgerDao(
-      |values({contract_id}, {party}, {ledger_offset}, {transaction_id})
-      |on conflict on constraint contract_divulgences_idx
-      |do nothing""".stripMargin
+    dbType match {
+      case Postgres =>
+        """insert into contract_divulgences(contract_id, party, ledger_offset, transaction_id)
+          |values({contract_id}, {party}, {ledger_offset}, {transaction_id})
+          |on conflict on constraint contract_divulgences_idx
+          |do nothing""".stripMargin
+      case H2Database =>
+        """insert into contract_divulgences(contract_id, party, ledger_offset, transaction_id)
+          |select {contract_id}, {party}, {ledger_offset}, {transaction_id}
+          |where not exists(select * from contract_divulgences where contract_id = {contract_id} and party = {party})
+          |""".stripMargin
+    }
 
   private val SQL_BATCH_INSERT_DIVULGENCES_FROM_TRANSACTION_ID =
-    """insert into contract_divulgences(contract_id, party, ledger_offset, transaction_id)
-      |select {contract_id}, {party}, ledger_offset, {transaction_id}
-      |from ledger_entries
-      |where transaction_id={transaction_id}
-      |on conflict on constraint contract_divulgences_idx
-      |do nothing""".stripMargin
+    dbType match {
+      case Postgres =>
+        """insert into contract_divulgences(contract_id, party, ledger_offset, transaction_id)
+          |select {contract_id}, {party}, ledger_offset, {transaction_id}
+          |from ledger_entries
+          |where transaction_id={transaction_id}
+          |on conflict on constraint contract_divulgences_idx
+          |do nothing""".stripMargin
+      case H2Database =>
+        """insert into contract_divulgences(contract_id, party, ledger_offset, transaction_id)
+          |select {contract_id}, {party}, ledger_offset, {transaction_id}
+          |from ledger_entries
+          |where transaction_id={transaction_id}
+          |and not exists(select * from contract_divulgences where contract_id = {contract_id} and party = {party})
+          |""".stripMargin
+    }
 
   private val SQL_INSERT_CHECKPOINT =
     SQL(
       "insert into ledger_entries(typ, ledger_offset, recorded_at) values('checkpoint', {ledger_offset}, {recorded_at})")
 
   private val SQL_IMPLICITLY_INSERT_PARTIES =
-    """insert into parties(party, explicit, ledger_offset)
-        |values({name}, {explicit}, {ledger_offset})
-        |on conflict (party)
-        |do nothing
-        |""".stripMargin
+    dbType match {
+      case Postgres =>
+        """insert into parties(party, explicit, ledger_offset)
+          |values({name}, {explicit}, {ledger_offset})
+          |on conflict (party)
+          |do nothing""".stripMargin
+      case H2Database =>
+        """insert into parties(party, explicit, ledger_offset)
+          |select {name}, {explicit}, {ledger_offset}
+          |where {name} not in (select party from parties)""".stripMargin
+    }
+
+  private val DUPLICATE_KEY_ERROR =
```

Comment:
I know this approach existed before this contribution, but it would be probably be a good idea to address this now: I'm not sure that relying on error messages is a very robust approach. Does H2 have some way of dealing with the errors that are dealt with using checks against this error string?

Reply:
As discussed at stand-up this is a bigger change, so for now I added a TODO.